File: app/src/main/java/com/nfcalarmclock/whatsnew/NacWhatsNewDialog.java
package com.nfcalarmclock.whatsnew;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.nfcalarmclock.R;
import com.nfcalarmclock.shared.NacSharedConstants;
import com.nfcalarmclock.util.dialog.NacDialogFragment;

/**
 */
public class NacWhatsNewDialog
	extends NacDialogFragment
{

	/**
	 * Listener for when what's new dialog has been read.
	 */
	@SuppressWarnings("UnnecessaryInterfaceModifier")
	public interface OnReadWhatsNewListener
	{
		public void onReadWhatsNew();
	}

	/**
	 * Tag for the class.
	 */
	public static final String TAG = "NacWhatsNewDialog";

	/**
	 * Listener for when a text-to-speech option and frequency is selected.
	 */
	private OnReadWhatsNewListener mOnReadWhatsNewListener;

	/**
	 * Call the OnReadWhatsNewListener object, if it has been set.
	 */
	public void callOnReadWhatsNewListener()
	{
		OnReadWhatsNewListener listener = this.getOnReadWhatsNewListener();

		if (listener != null)
		{
			listener.onReadWhatsNew();
		}
	}

	/**
	 * Get the OnReadWhatsNewListener object.
	 *
	 * @return The OnReadWhatsNewListener object.
	 */
	public OnReadWhatsNewListener getOnReadWhatsNewListener()
	{
		return this.mOnReadWhatsNewListener;
	}

	/**
	 * Called when the dialog is canceled.
	 */
	@Override
	public void onCancel(DialogInterface dialog)
	{
		this.callOnReadWhatsNewListener();
	}

	/**
	 * Called when the dialog is created.
	 */
	@NonNull
	@Override
	public Dialog onCreateDialog(@Nullable Bundle savedInstanceState)
	{
		this.setupSharedPreferences();

		NacSharedConstants cons = this.getSharedConstants();

		return new AlertDialog.Builder(requireContext())
			.setTitle(cons.getTitleWhatsNew())
			.setPositiveButton(cons.getActionOk(), (dialog, which) ->
				this.callOnReadWhatsNewListener())
			.setView(R.layout.dlg_whats_new)
			.create();
	}

	/**
	 */
	@Override
	public void onResume()
	{
		super.onResume();

		// Initialize the widgets
		AlertDialog dialog = (AlertDialog) getDialog();
		TextView textview = dialog.findViewById(R.id.whats_new_version);

		// Prepare the strings
		NacSharedConstants cons = this.getSharedConstants();
		String versionName = textview.getText().toString();
		String versionNum = cons.getAppVersion();

		// Set the version
		textview.setText(versionName + " " + versionNum);
	}

	/**
	 * Set the OnReadWhatsNewListener object.
	 *
	 * @param  listener  The OnReadWhatsNewListener object.
	 */
	public void setOnReadWhatsNewListener(OnReadWhatsNewListener listener)
	{
		this.mOnReadWhatsNewListener = listener;
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/whatsnew/NacWhatsNewDialog.kt
package com.nfcalarmclock.whatsnew

import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.FragmentManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.button.MaterialButton
import com.nfcalarmclock.R
import com.nfcalarmclock.view.dialog.NacBottomSheetDialogFragment
import com.nfcalarmclock.view.toThemedBold
import com.nfcalarmclock.whatsnew.NacWhatsNewDialog.OnReadWhatsNewListener

/**
 * Show what is new with the app after an update.
 */
class NacWhatsNewDialog
	: NacBottomSheetDialogFragment()
{

	/**
	 * Listener for when what's new dialog has been read.
	 */
	fun interface OnReadWhatsNewListener
	{
		fun onReadWhatsNew()
	}

	/**
	 * Listener for when a text-to-speech option and frequency is selected.
	 */
	var onReadWhatsNewListener: OnReadWhatsNewListener? = null

	/**
	 * Called when the dialog is canceled.
	 */
	override fun onCancel(dialog: DialogInterface)
	{
		// Call the listener
		onReadWhatsNewListener?.onReadWhatsNew()
	}

	/**
	 * Called when the creating the view.
	 */
	override fun onCreateView(
		inflater: LayoutInflater,
		container: ViewGroup?,
		savedInstanceState: Bundle?
	): View?
	{
		return inflater.inflate(R.layout.dlg_whats_new, container, false)
	}

	/**
	 * Called when the view has been created.
	 */
	override fun onViewCreated(view: View, savedInstanceState: Bundle?)
	{
		// Super
		super.onViewCreated(view, savedInstanceState)

		// Get the theme color
		val themeColor = sharedPreferences.themeColor

		// Get all the whats new items and add theme color to bold sections
		val allMessages = resources.getTextArray(R.array.whats_new_items)
			.map { it.toString().toThemedBold(themeColor) }
			.toList()

		// Setup the views
		setupOkButton()
		setupRecyclerView(allMessages)
		setBehaviorState(BottomSheetBehavior.STATE_EXPANDED)
	}

	/**
	 * Setup the OK button.
	 */
	private fun setupOkButton()
	{
		// Get the ok button
		val okButton: MaterialButton = dialog!!.findViewById(R.id.ok_button)

		// Setup the ok button
		setupPrimaryButton(okButton, listener = {

			// Call the listener
			onReadWhatsNewListener?.onReadWhatsNew()

			// Dismiss
			dismiss()

		})
	}

	/**
	 * Setup the recycler view.
	 */
	private fun setupRecyclerView(allMessages: List<String>)
	{
		// Get the views
		val recyclerView = dialog!!.findViewById<RecyclerView>(R.id.whats_new_recyclerview)
		val listAdapter = NacWhatsNewListAdapter()

		// Setup the list adapter
		listAdapter.submitList(allMessages)

		// Setup the recyclerview
		recyclerView.adapter = listAdapter
		recyclerView.layoutManager = LinearLayoutManager(context)

		// Set the height of the recyclerview
		setupScrollableViewHeight(recyclerView, 60)
	}

	companion object
	{

		/**
		 * Tag for the class.
		 */
		const val TAG = "NacWhatsNewDialog"

		/**
		 * Show the dialog.
		 */
		fun show(
			manager: FragmentManager,
			listener: () -> Unit = {})
		{
			// Create the dialog
			val dialog = NacWhatsNewDialog()

			// Setup the listener
			dialog.onReadWhatsNewListener = OnReadWhatsNewListener {
				listener()
			}

			// Show the dialog
			dialog.show(manager, TAG)
		}

	}

}